pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -886,43 +932,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
 	return volume, err
 }
 
-func (os *OpenStack) getComputeIDbyHostname(cClient *gophercloud.ServiceClient) (string, error) {
+// getLocalComputeID tries to find ID of the OpenStack instance where
+// this code runs.
+func (os *OpenStack) getLocalComputeID(cClient *gophercloud.ServiceClient) (string, error) {
+	// try to read /var/lib/cloud/data/instance-id
+	const instanceIDFile = "/var/lib/cloud/data/instance-id"
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
+		id := string(idBytes)
+		id = strings.TrimSpace(id)
+		glog.V(5).Infof("Got instance id from %s: %s", instanceIDFile, id)
+		return id, nil
+	}
+	glog.V(4).Infof("Reading %s failed: %v, trying hostname", instanceIDFile, err)
 
+	// try to find ID from hostname
```

Comment:
Ideally we'd never need to guess instance-id from hostname.  What would it take to remove this fallback?

The other methods that might be provided in OpenStack (afaik) are:
- metadata service
- config drive

[Metadata service](http://docs.openstack.org/admin-guide-cloud/compute-networking-nova.html#metadata-service) is easy:  Just fetch http://169.254.169.254/openstack/2012-08-10/meta_data.json and return the `uuid` value

[Config drive](http://docs.openstack.org/user-guide/cli_config_drive.html) is more complicated - we need to mount the block device labelled `config-2` and then parse the same `uuid` value from `<mount point>/openstack/2012-08-10/meta_data.json`.  This obviously requires the permissions to mount filesystems - but the kubelet already has that for persistent volumes (always?).

Regardless of the method used to fetch it, we should cache value in memory since it will never change and will be used multiple times by the other functions in this file.